### PR TITLE
Add support for branches in the travis integration

### PR DESCRIPTION
**Pull request by gardleopard**
You can choose to listen to org/repo for all builds of the project, or you can choose a specific branch like org/repo/branch.

I have checked how jenkins work with https://wiki.jenkins-ci.org/display/JENKINS/Multi-Branch+Project+Plugin and this gives almost the same experience.

### CODE REVIEW

**Comment by gardleopard** [comment]
@spinnaker/reviewers PTAL

**Comment by gardleopard** [comment]
I have added a virtual branch called tags, if the build is a tag, it will report it as branch tags. So now this integration lets you create pipeline that subscribes to:
* Every build of the project
* Every build of a specific branch
* Every build of a tag (through the virtual branch tags)

This should make the travis integration very flexible when choosing the workflow the user wants for a specific project.

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/TravisBuildMonitor.groovy [hunk 1]
```diff
@@ -161,13 +168,14 @@ class TravisBuildMonitor implements PollingMonitor{
...
                             log.info "sending build events for builds between ${lastBuild} and ${currentBuild}"
                             for (int buildNumber = lastBuild+1; buildNumber < currentBuild; buildNumber++) {
-                                Build build = buildMasters.map[master].getBuild(repo, buildNumber) //rewrite to afterNumber list thing
-                                log.info "pushing event for ${master}:${repo.slug}:${build.number}"
-                                String url = "${buildMasters.map[master].baseUrl}/${repo.slug}/builds/${build.id}"
-                                GenericProject project = new GenericProject(repo.slug, new GenericBuild((build.state == BUILD_IN_PROGRESS), build.number, build.duration, TravisResultConverter.getResultFromTravisState(build.state), repo.slug, url))
-
-                                echoService.postEvent(
-                                    new GenericBuildEvent(content: new GenericBuildContent(project: project, master: master, type: 'travis')))
+                                Build build = travisService.getBuild(repo, buildNumber) //rewrite to afterNumber list thing
```
Comment (lwander) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/TravisBuildMonitor.groovy]: Can build be null? Maybe `build?.state` below.

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/TravisCache.groovy [hunk 2]
```diff
@@ -58,7 +58,7 @@ class TravisCache {
 
     int setQueuedJob(String master, String jobName, int buildNumber) {
         Jedis resource = jedisPool.getResource()
-        int queueId = (int) (long) new Date().getTime()
+        int queueId = (int) (long) new Date().getTime() / 1000
```
Comment (lwander) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/TravisCache.groovy]: Use `TimeUnit.MILLISECONDS.fromSeconds(new Date.getTime())`

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/service/TravisService.groovy [hunk 3]
```diff
@@ -220,8 +236,34 @@ class TravisService implements BuildService {
...
+    }
+
     @Override
     BuildServiceProvider buildServiceProvider() {
         return BuildServiceProvider.TRAVIS
     }
+
+    protected String cleanRepoSlug(String inputRepoSlug) {
+        def parts = inputRepoSlug.tokenize('/')
+        return "${parts[0]}/${parts[1]}"
```
Comment (lwander) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/service/TravisService.groovy]: Are you sure parts will have 2 components?

Comment (gardleopard) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/service/TravisService.groovy]: Im not sure, I think all github and stash repos has org/repo but I am not 100% sure.

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/TravisCache.groovy (hunk 2 above)
Comment (lwander) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/TravisCache.groovy]: Also are you sure that seconds offers enough granularity?

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/TravisBuildMonitor.groovy (hunk 1 above)
Comment (gardleopard) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/TravisBuildMonitor.groovy]: @lwander I can add that, but from my tests build did not become null when it was missing. build.number was 0 and the other fields null.

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/client/model/Commit.groovy [hunk 4]
```diff
@@ -37,7 +37,24 @@ class Commit {
...
+    boolean isTag(){
+        if (!compareUrl) {
+            return false
+        }
+        return compareUrl.split("/compare/").last().matches(branch)
+    }
+
+    String branchNameWithTagHandling() {
+        if (isTag()) {
+            return "tags"
```
Comment (lwander) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/client/model/Commit.groovy]: What if someone has a branch named `"tags"`?

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/client/model/Commit.groovy [hunk 5]
```diff
@@ -37,7 +37,24 @@ class Commit {
     @SerializedName("author_name")
     String authorName
 
+    @SerializedName("compare_url")
+    String compareUrl
```
Comment (lwander) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/client/model/Commit.groovy]: Is this a github specific thing? I think this `Commit` is supposed to be agnostic to the git host

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/TravisBuildMonitor.groovy (hunk 1 above)
Comment (lwander) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/TravisBuildMonitor.groovy]: Ah that should be fine then like it is

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/client/model/Commit.groovy (hunk 4 above)
Comment (gardleopard) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/client/model/Commit.groovy]: Then tagged builds will come as events on the same branch.

File: igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/client/model/Commit.groovy (hunk 5 above)
Comment (gardleopard) [inline igor-web/src/main/groovy/com/netflix/spinnaker/igor/travis/client/model/Commit.groovy]: I think travis is github specific. https://docs.travis-ci.com/user/customizing-the-build/#What-git-Repository-Providers-can-I-use
